### PR TITLE
Add 'keep_firing_for' field to alerting rules

### PR DESCRIPTION
**Pull request by roidelapluie**
This commit adds a new 'keep_firing_for' field to Prometheus alerting rules. The 'keep_firing_for' field specifies the minimum amount of time that an alert should remain firing, even if the expression does not return any results.

This feature was discussed at a previous dev summit, and it was determined that a feature like this would be useful in order to allow the expression time to stabilize and prevent confusing resolved messages from being propagated through Alertmanager.

This approach is simpler than having two PromQL queries, as was sometimes discussed, and it should be easier to implement.

This commit does not include tests for the 'resolve_delay' field.  This is intentional, as the purpose of this commit is to gather comments on the proposed design of the 'resolve_delay' field before implementing tests. Once the design of the 'resolve_delay' field has been finalized, a follow-up commit will be submitted with tests.

See https://github.com/prometheus/prometheus/issues/11570

Signed-off-by: Julien Pivotto <roidelapluie@o11y.eu>

### CODE REVIEW

**Comment by juliusv** [comment]
Does it (upon resolution) just delay when a resolved notification gets sent out, or does it actually prevent the alert itself from getting resolved across a bridge of data-less time, so that it continues to fire even after the delay if the data reappears, and it doesn't have to wait the `for` duration again?

**Comment by roidelapluie** [comment]
It delays when a resolved notification gets sent out, but it also prevents the alert itself from getting resolved across a bridge of data-less time. The alert will continue to fire even after the delay if the data reappears, and it does not have to wait for the "for" duration again. This feature allows the alert to stay active until the system truly stabilizes, it's important to note that it doesn't add more time to the for duration.

**Comment by juliusv** [comment]
Great, yeah that sounds like the good way of those two options.

I see you already went through multiple naming options, `resolve_delay` vs. `keep_firing_for` in the code itself now. `resolve_delay` might even have been a tad clearer, because `keep_firing_for` could also be read as just a minimum firing duration with all data being there, and once that's passed you would immediately resolve when a series disappears.

Since it's a quite fundamental Prometheus feature, it'd be nice to get input from one or two more team members on the naming and exact semantics though.

**Comment by roidelapluie** [comment]
cc @beorn7 @LeviHarrison @codesome

**Comment by beorn7** [comment]
At first glance, I like it. As an "inverted `for`", it seems to fill a logical gap.

I should note, though, that it addresses a distinctly different use case from the "hysteresis" one, even if both are in the area of "stabilizing" alert resolution.

Looking forward to the thoughts of others.

**Comment by beorn7** [comment]
Minor UI idea: Maybe we should display the state "still firing although the alert condition isn't true anymore" separately in the UI? We display "pending" (alert condition applies, but not for long enough yet) and "firing" (alert condition has applied for long enough), so the "inverted pending" (alert condition doesn't apply anymore , but not for long enough yet) should perhaps be called out explicitly, too.

File: rules/alerting.go [hunk 1]
```diff
@@ -404,16 +415,29 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
 	// Check if any pending alerts should be removed or fire now. Write out alert timeseries.
 	for fp, a := range r.active {
 		if _, ok := resultFPs[fp]; !ok {
+			var keepFiring bool
```
Comment (bboreham) [inline rules/alerting.go:418]: I would find it useful to have a comment hereabouts saying something like "no results for this fingerprint; the alert is no longer firing".

File: rules/alerting.go [hunk 2]
```diff
@@ -404,16 +415,29 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
...
-			if a.State != StateInactive {
+			if a.State != StateInactive && !keepFiring {
 				a.State = StateInactive
 				a.ResolvedAt = ts
 			}
-			continue
+			if !keepFiring {
+				continue
+			}
+		} else {
```
Comment (bboreham) [inline rules/alerting.go:439]: I would find it useful to have a comment hereabouts saying something like "the alert is firing".

**Review by bboreham — APPROVED** [review]
Seems fine; I had a little trouble following the if-else logic.

One reaction was that the default should be greater than zero, e.g. it could be the stale interval.
Or perhaps the user could have an option to set a system-wide default?

**Comment by bboreham** [comment]
Please can you clarify how the two terms 'keep_firing_for' and 'resolve_delay' are used in this PR.
AFAICS the latter is only in the title and description.

**Comment by roidelapluie** [comment]
> Please can you clarify how the two terms 'keep_firing_for' and 'resolve_delay' are used in this PR. AFAICS the latter is only in the title and description.

I started using resolve_delay but ended up calling it keep_firing_for

**Comment by juliusv** [comment]
:+1: From my side. I was still trying to think of an even better name for the setting, but I didn't come with anything that was clearly better in the end :)

**Comment by grobinson-grafana** [comment]
I too like `KeepFiringFor`, some other options I thought of could be `MinFiringDuration` and `ResolveAfter`, but `ResolveAfter` is arguably ambiguous because you also need to wait for data to be absent.

**Comment by beorn7** [comment]
I like `keep_firing_for`. It includes the `for`, af which it is kind of the inverse. And we do use the term "firing" elsewhere.

**Comment by beorn7** [comment]
Note my thoughts expressed in https://github.com/prometheus/prometheus/issues/10536#issuecomment-1385509124 , or in other words: I think this PR makes sense as-is, but we might still want to consider a "real" hysteresis implementation for later (and both might even combine).

**Comment by roidelapluie** [comment]
Thanks all!

**Comment by calestyo** [comment]
Hey.

Would be interested to hear what the people who developed `keep_firing_for` think about my alternative (or I should better say "additive" - as it's not mean to replace the current `keep_firing_for`) proposal I've made in #14085.

Cheers,
Chris

**Pull request merged**